Day_3/day_3.py

```diff
@@ -0,0 +1,428 @@
+import os
+import numpy as np
+
+import math
+
+class Box:
+    def __init__(self, box_length, particles=None):
+        self.box_length = box_length
+        self.particles = particles # if None, no particles in MCSystem
+        
+        # Calculated quantities
+        self.calculate_number_particles()
+    
+    def calculate_number_particles(self):
+        """Calculate the number of particles"""
+        if self.particles is None:
+            self.number_particles = 0
+        else:
+            self.number_particles = len(self.particles)
```

Comment:
Can you explain more what you mean here? If you have an if statement like `if self.particles` (which is in your alternative solution) and `self.particles` is a `numpy` array, you will get a value error (ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all() )

`a.any()` or `a.all()` is not applicable because we are checking that it is a numpy array (ie that it set and is not None). An alternative to would be to use `type` or `isinstance`, I believe. However, I don't understand why the proposed solution is bad practice.

Reply:
The issue is the `self.number_particles` is a computed out of `self.particles`. And later when you use `self.number_particles`  directly in can get out of sync with the actual length of the `self.particles`. This is called the data integrity issue. The proposed way is to get `len(self.particles)` every time you want to use it. This is a good practice to teach the students.

I didn't know you are using numpy arrays for `self.particles`, if so, then modify it to work with numpy, but without a pre-calculated `self.number_particles`

Reply:
Got it. Could we use `@property` here similar to how it is used in the `Box` class for `volume`?

Reply:
Yes, sure, this would be good.